opentelemetry-instrumentation/src/opentelemetry/instrumentation/metric.py

```diff
@@ -0,0 +1,74 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# type: ignore
+
+"""
+OpenTelemetry Instrumentation Metric mixin
+"""
+from contextlib import contextmanager
+from time import time
+
+from opentelemetry import metrics
+from opentelemetry.sdk.metrics import PushController, ValueRecorder
+from opentelemetry.trace import SpanKind
+
+
+class MetricMixin:
+    """Used to record metrics related to instrumentations."""
+
+    def init_metrics(self, name, version, exporter=None, interval=None):
+        self._meter = metrics.get_meter(name, version)
+        if exporter and interval:
+            self._controller = PushController(
+                meter=self._meter, exporter=exporter, interval=interval
+            )
```

Comment:
I was actually thinking about this approach originally, just expose the `meter` so users can decide what they want to do with it. I'm okay with t removing the automatically started pipeline. However, this feature requires the dependency on `opentelemetry-sdk` regardless, because of line 63 in `metric.py` in `opentelemetry-instrumentation`. To enable this, the sdk implementation of `ValueRecorder` must be used, so we MUST have a dependency on it. I think the message is: if you want to autocollect these metrics, this will be a feature offered by the default sdk and so you have to install it. I also don't know if the api and sdk separation of metrics makes sense today. Will we ever even have a different implementation of the metrics sdk? It's not really the same as the tracer, in which we would.

If there is a big push to not take a dependency on the sdk, we would probably have to change our metrics api implementation and create an explicit `create_value_recorder` method for the `meter` s. I'm actually okay with this, but might want to leave this for another pr. Thoughts?

Reply:
> If there is a big push to not take a dependency on the sdk, we would probably have to change our metrics api implementation and create an explicit `create_value_recorder` method for the `meter` s. I'm actually okay with this, but might want to leave this for another pr. Thoughts?

Ya I agree it would be easy to work around that use of `ValueRecorder` later if we decide the instrumentation shouldn't take dependency on the SDK (it would be breaking change to the Instrumentor constructor though?). Could you explain why the api/sdk separation isn't the same with metrics as it is with tracer (I think I'm missing some background info)?

Would be great to get other Python SIG folks' thoughts on this too

Reply:
> Could you explain why the api/sdk separation isn't the same with metrics as it is with tracer

I feel like we don't expect people to implement their own sdk to record + export metrics. It's as if we introduced the api + sdk separation simply to match what trace is doing.

Reply:
This discussion makes me wonder if the behaviour exposed in the ValueRecorder by the SDK should just be in the API then, instead of having an interface that is not usable with the SDK. If this isn't already being discussed in the metrics SIG, it should, I would suspect other SIGs are running into similar issues.

Reply:
[Java](https://github.com/open-telemetry/opentelemetry-go/blob/7d718676444c9c2f18f8355ae262b64744fed331/api/metric/meter.go#L111) implementation and [Go](https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/MeterSdk.java#L62) implementation both have constructors for each metric instrument. So depending on whether they are using the sdk `MeterProvider` or not, it returns different implementations of the instrument, so they have an interface that is useable by the SDK.